lib/ddtrace/profiling/ext/forking.rb

```diff
@@ -37,20 +36,45 @@ module Kernel
           FORK_STAGES = [:prepare, :parent, :child].freeze
 
           def fork
-            wrapped_block = proc do
-              # Trigger :child callback
-              at_fork_blocks[:child].each(&:call) if at_fork_blocks.key?(:child)
-              yield
-            end
+            # If a block is provided, it must be wrapped to trigger callbacks.
+            child_block = if block_given?
+                            proc do
+                              # Update current thread clock, if available.
+                              if Thread.current.respond_to?(:update_native_ids, true)
+                                Thread.current.send(:update_native_ids)
+                              end
+
+                              # Trigger :child callback
+                              at_fork_blocks[:child].each(&:call) if at_fork_blocks.key?(:child)
+                              yield
+                            end
+                          end
 
             # Trigger :prepare callback
             at_fork_blocks[:prepare].each(&:call) if at_fork_blocks.key?(:prepare)
 
             # Start fork
-            result = super(&wrapped_block)
+            # If a block is provided, use the wrapped version.
+            result = child_block.nil? ? super : super(&child_block)
+
+            # Trigger correct callbacks depending on whether we're in the parent or child.
+            # If we're in the fork, result = nil: trigger child callbacks.
+            # If we're in the parent, result = fork PID: trigger parent callbacks.
+            # rubocop:disable Style/IfInsideElse
+            if result.nil?
+              # Update current thread clock, if available.
+              if Thread.current.respond_to?(:update_native_ids, true)
```

Comment:
Also worth noting that if we don't manually update native IDs here after forking, there's no practical way to get CPU profiling working on the thread that forked.

This is because clock IDs on threads change after a fork, and the only way you can update the clock ID for a thread is by getting the clock ID from *within* the execution context of the thread in question. While our `Thread` patch covers background threads via patching `Thread#initialize`, that strategy does not cover the thread that causes the fork, because it will not reinitialize.

Since the thread that forks calls this function, from within the execution context of the thread, it seems fitting to update the IDs for it as we trigger `at_fork` callbacks too.

A downside to this is it creates a soft coupling between this fork patch and the implementation of our CPU time patch for Thread. But I think it's worthwhile until we have a better solution that solves the same problem.

Reply:
> A downside to this is it creates a soft coupling between this fork patch and the implementation of our CPU time patch for Thread. But I think it's worthwhile until we have a better solution that solves the same problem.

Could this be made into a `at_fork_blocks[:child]` callback proc?

Reply:
I think the answer is yes. I thought about doing this before, but I guess I forgot to implement it. This would be the place to do it:

Reply:
I guess as an amendment to the above, it doesn't have to go in that function, it could go in another function in the same `Setup` class. So maybe this is a sensible place to do it.

Reply:
I see that "CPU timing/native IDs" are not strictly tied to profiling, but they are effectively only used by profiling today.

I do acknowledge that they are not at the same level "activate the profiler" in terms of responsibility alignment.

I think that any semantic separation would suffice here: a separate method, like you suggested, or a separate callback would be enough of a separation in my option.

Reply:
Alright, I implemented this change.